diff --git a/Setup.hs b/Setup.hs
index fabdd91..9a994af 100644
--- a/Setup.hs
+++ b/Setup.hs
@@ -1,14 +1,2 @@
 import Distribution.Simple
-import Distribution.Verbosity
-import Distribution.Simple.Utils
-
-prepareBuild =
-    rawSystemExit normal "git" ["submodule", "update", "--init", "--recursive"]
-
-main =
-    defaultMainWithHooks simpleUserHooks
-    { preBuild = \args buildFlags ->
-          do
-              prepareBuild
-              preBuild simpleUserHooks args buildFlags
-    }
+main = defaultMain
diff --git a/bindings-freetype-gl.cabal b/bindings-freetype-gl.cabal
index e960316..f7121ec 100644
--- a/bindings-freetype-gl.cabal
+++ b/bindings-freetype-gl.cabal
@@ -10,7 +10,7 @@ author:              Eyal Lotem
 maintainer:          eyal.lotem@gmail.com
 copyright:           2016 Eyal Lotem
 category:            Graphics, Fonts, Bindings
-build-type:          Custom
+build-type:          Simple
 extra-source-files:  windows/glew.c
                      windows/GL/glew.h
                      windows/GL/wglew.h
@@ -26,11 +26,6 @@ extra-source-files:  windows/glew.c
                      freetype-gl/vertex-attribute.h
                      freetype-gl/vertex-buffer.h
 
-custom-setup
-  setup-depends:
-                 base  >= 4.5
-               , Cabal >= 2.2
-
 library
   exposed-modules:     Bindings.FreetypeGL.FontManager
                      , Bindings.FreetypeGL.Markup
