diff --git a/src/Language/Nix/Identifier.hs b/src/Language/Nix/Identifier.hs
index 0d7e5ba..e457a7d 100644
--- a/src/Language/Nix/Identifier.hs
+++ b/src/Language/Nix/Identifier.hs
@@ -13,11 +13,11 @@ import Control.DeepSeq
 import Control.Lens
 import Data.Char
 import Data.String
-import Distribution.Compat.ReadP as ReadP
 import Distribution.Text
 import GHC.Generics ( Generic )
 import Prelude.Compat
 import Test.QuickCheck
+import Text.ParserCombinators.ReadP as ReadP
 import Text.PrettyPrint as PP
 
 -- | Identifiers in Nix are essentially strings. They can be constructed
@@ -53,9 +53,9 @@ instance Arbitrary Identifier where
   arbitrary = Identifier <$> arbitrary
   shrink (Identifier i) = map Identifier (shrink i)
 
-instance Text Identifier where
-  disp = view (ident . to quote . to text)
-  parse = parseQuotedIdentifier <++ parseSimpleIdentifier
+--instance Text Identifier where
+--  disp = view (ident . to quote . to text)
+--  parse = parseQuotedIdentifier <++ parseSimpleIdentifier
 
 -- | 'ReadP' parser for simple identifiers, i.e. those that don't need
 -- quoting.
