diff --git b/Setup.hs a/Setup.hs
index fad5026..cf05462 100644
--- b/Setup.hs
+++ a/Setup.hs
@@ -1,3 +1,5 @@
+{-# LANGUAGE CPP #-}
+
 import Distribution.Simple
 import Distribution.Simple.Setup
 import Distribution.PackageDescription
@@ -26,7 +28,11 @@ main = defaultMainWithHooks simpleUserHooks {
   where
     defaultPostConf :: Args -> ConfigFlags -> PackageDescription -> LocalBuildInfo -> IO ()
     defaultPostConf args flags pkgdescr lbi = do
+#if MIN_VERSION_Cabal(2,3,0)
+      libdir_ <- getDbProgramOutput (fromFlag (configVerbosity flags))
+#else
       libdir_ <- rawSystemProgramStdoutConf (fromFlag (configVerbosity flags))
+#endif
                      ghcProgram (withPrograms lbi) ["--print-libdir"]
       let libdir = reverse $ dropWhile isSpace $ reverse libdir_
 
