diff --git a/Basement/Block.hs b/Basement/Block.hs
index 07c8c3b..ff7c399 100644
--- a/Basement/Block.hs
+++ b/Basement/Block.hs
@@ -66,6 +66,7 @@ module Basement.Block
     , withPtr
     ) where
 
+import           Control.Monad
 import           GHC.Prim
 import           GHC.Types
 import           GHC.ST
diff --git a/Basement/Block/Mutable.hs b/Basement/Block/Mutable.hs
index 0710c92..1be83f6 100644
--- a/Basement/Block/Mutable.hs
+++ b/Basement/Block/Mutable.hs
@@ -64,6 +64,7 @@ module Basement.Block.Mutable
     , copyToPtr
     ) where
 
+import           Control.Monad
 import           GHC.Prim
 import           GHC.Types
 import           Basement.Compat.Base
diff --git a/Basement/BoxedArray.hs b/Basement/BoxedArray.hs
index 4fb7d1b..c6a874f 100644
--- a/Basement/BoxedArray.hs
+++ b/Basement/BoxedArray.hs
@@ -74,6 +74,7 @@ module Basement.BoxedArray
     , builderBuild_
     ) where
 
+import           Control.Monad
 import           GHC.Prim
 import           GHC.Types
 import           GHC.ST
diff --git a/Basement/String.hs b/Basement/String.hs
index 98165e6..5c0b6ea 100644
--- a/Basement/String.hs
+++ b/Basement/String.hs
@@ -95,6 +95,7 @@ module Basement.String
     , toBase64OpenBSD
     ) where
 
+import           Control.Monad
 import           Basement.UArray           (UArray)
 import qualified Basement.UArray           as Vec
 import qualified Basement.UArray           as C
diff --git a/Basement/UArray.hs b/Basement/UArray.hs
index 55f69ed..a2ca3b7 100644
--- a/Basement/UArray.hs
+++ b/Basement/UArray.hs
@@ -106,6 +106,7 @@ module Basement.UArray
     , toBase64Internal
     ) where
 
+import           Control.Monad
 import           GHC.Prim
 import           GHC.Types
 import           GHC.Word
diff --git a/Basement/UArray/Base.hs b/Basement/UArray/Base.hs
index d766396..15225e1 100644
--- a/Basement/UArray/Base.hs
+++ b/Basement/UArray/Base.hs
@@ -53,6 +53,7 @@ module Basement.UArray.Base
     , pureST
     ) where
 
+import           Control.Monad
 import           GHC.Prim
 import           GHC.Types
 import           GHC.Ptr
diff --git a/Basement/UTF8/Base.hs b/Basement/UTF8/Base.hs
index cc4fb26..7f7b28c 100644
--- a/Basement/UTF8/Base.hs
+++ b/Basement/UTF8/Base.hs
@@ -15,6 +15,7 @@
 module Basement.UTF8.Base
     where
 
+import           Control.Monad
 import           GHC.ST (ST, runST)
 import           GHC.Types
 import           GHC.Word
